### ISSUE TITLE
Click-repl pkg tab-completion of option values fails - need new click-repl version with fix

### ISSUE THREAD

**Issue by KSchopmeyer** [pywbem/pywbemtools#1279]
Click-repl package version 0.2.0 tab-completion of option values (i.e. in the repl mode) fails.  The package now has a fix for this issue https://github.com/click-contrib/click-repl/pull/88 which has been committed to the master but it has not been released.  For us this means that the tab-completion for options like --name, --mock-server, --connection-file, etc. will not work in repl mode but will work in command line mode.

However, the changes for this new release (date unknown) also removes support for python 2.

The alternatives are probably to:

1. localize that package based on its current master until they do a release (ex. fork it). I recommend we not do this.
2. Document the issue and note to users that the repl tab-completion of these option/argument values does not work in this release.

Proposal: document the deficiency in pywbemcli docs but not try to use their master branch as our package version.

Issue noted in documentation for tab-completion in pr #1282, (generalization of tab-completion documentation) as a note.

Note that there is a second issue with click-repl in that the tab completion does not include tab-completion for the click.Path() type which is part of tab-completion for all of the shells which click-tab-completion supports.  This is documented in click-repl issues in issue https://github.com/click-contrib/click-repl/issues/59 but since that issue is several years old, it would appear that nobody is taking responsibility for a fix.

**Comment by andy-maier** [comment]
Do the repl users really use tab completion on options that change the target server?

I think that we should use the released version.

**Comment by KSchopmeyer** [comment]
We concluded that we will note this restriction in the docs and the changes.rst.

**Comment by KSchopmeyer** [comment]
The note about the restriction and issues with python 2.7 and python 3.5 was included in PR #1247 and PR #1278.  I proposed that this issue be held open but not tied to pywbemtools  1.2.0 as it is dependent on the release of the next version of click-repl.

**DISCUSSION**

**Comment by GhostOps77** [comment]
So, do u want the python 2 support for the click-repl package?

**Comment by KSchopmeyer** [comment]
We still support python 2 but since the whole premise of tab-complete/autocomplete is different between click 8 and click 7 and we think that the primary users for python 2 are some older test environments, we are comfortable for python2 with not including the completion of argument and option values for python 2. Therefore since the current click-repl version is python 3 only, we define the old version for python 2 and python 3.5 as part of the requirements.  We are probably better off not worrying about tab-completion for args and options with click 7 because that would lead to other differences between the shell tab-completion and repl tab complation (the click at option/argument shell_completion, etc.

Actually what we really want is the release of a new click-repl that includes click-repl pr https://github.com/click-contrib/click-repl/pull/88 the extension to handle the argument and option values in the repl mode.   Right now we put a note in the documentation that argument/option value tab-completion does not work in repl mode for this release.  Note that since we have not set a maximum click-repl version in requirements, if a new release is made, the args/option value tab-completion in our code should work (sort of a surprise for the user).  

Anything I can to to speed up the next click-repl release I will be glad to help with.  NOTE: I did run a test of the click-repl master in our pywbemtools test suite and I think I hit a failure but I will have to rerun the test again.

Thanks for asking.

**Comment by GhostOps77** [comment]
> Thanks for asking the question above.
> 
> We still support python 2/3.5 but since the whole premise of tab-complete/autocomplete is different between click 8 and click 7 and we think that the primary users for python 2 are some older autoated test environments, not our interactive users, we are more comfortable with python 2 not including the completion of argument and option values. Therefore since the current click-repl version is python 3 only, we define the old click-repl version for python 2 and python 3.5 as part of install requirements. We are probably better off not worrying about tab-completion for args and options with click 7 because that would lead to other differences between the shell tab-completion and repl tab completion (the click at option/argument shell_completion, etc.
> 
> Actually what we really would like is the release of a new click-repl version that includes click-repl pr [click-contrib/click-repl#88](https://github.com/click-contrib/click-repl/pull/88) (the extension to handle the argument and option values) in the repl mode ASAP. Right now we put a note in our documentation that argument/option value tab-completion does not work in repl mode for this release. Note that since we have not set a maximum click-repl version in requirements, if a new release is made, the args/option value tab-completion in our code should work (sort of a surprise for the user).
> 
> Anything I can to to speed up the next click-repl release I will be glad to help with.
> 
> NOTE: I did run a test of the click-repl master in our pywbemtools test suite and also manually in repl mode and I think I hit a failure with processing click.path options which works in the shell but not in the interactive mode( see below). I have not filed a bug because I want to create a simple example to demo the issue. NOTE: We do not have functional tests for tab-completion from the command line or in interactive mode in our test suite yet. We just have unit tests of our own shell completer functions so the pywbemtools automated tests do not pick this up.
> 
> Example of the issue I a seeing now:
> 
> ```
>   With an option that has the type click.Path() and no shell_completion (i.e, pywbemcli --connections-file)
> 
>    In the bash shell:
> 
>     pywbemcli -C /home/kschopmeyer/.pywbemcli_connect<TAB>
>     produces:
>     pywbemcli -C /home/kschopmeyer/.pywbemcli_connections.yaml
> 
> In the repl mode (with the clic-repl master branch):
>     pywbmcli$  -C /home/kschopmeyer/.pywbemcli_connect<TAB>
>    does nothing
> 
> at the same time:
> With an option that is type str and includes the shell_complete attribute:
> 
> in both the repl mode and command mode, the tab-completion works:
>    
>    pywbemcli$  -n mock3n<TAB>
>    produces:
>                        -n mock3ns        <----which is correct.
> ```

yeah, i understand it why, cuz we still haven't implemented the autocompletion for filenames and filepath

Will implement it soon, and ping u later. thanks for the reply

**Comment by KSchopmeyer** [comment]
ACTION: What we will need to do in the future is to require the new version of click-repl when it is released.

Karl needs to write an issue in click-repl documenting in simple example the problem with tab-completion master today.

**Comment by GhostOps77** [comment]
@KSchopmeyer I'm sorry, but I was busy for the past one week
that's why I was unable to resume the feature updates in click-repl

I'll start to add some changes to the click-repl from now on
Currently, I'm updating all the changes in [my click-repl fork](https://github.com/GhostOps77/click-repl/tree/GhostOps77-patch-1)

And also as I said, It would be a great if you were able to contribute changes to the click-repl module

And also, I'm having a very big trouble on making the module as python 2 compatible
So, can you give me a reason that why you are making your module to support for python 2?

**Comment by andy-maier** [comment]
MacOS no longer ships Python 2.7 as of its version 12.3, and major Linux distributions have removed it. However, googling "install python 2.7 on linux" has 129 million hits, which tells you something.

Github Actions supports running Python 2.7 even on ubuntu-latest which by now is 22.04 which no longer includes it by default, so the Github folks reinstall it even there.

So we have to realize that people are using Python 2.7 longer than anticipated.

In addition, the pywbem package is included as a package in RHEL. RHEL is pretty much the most conservative Linux distro ni terms of supporting old versions of stuff. RedHat supports Python 2.7 in RHEL 8 until 6/2024, and has removed it from RHEL 9. Unfortunately, the RedHat package maintainers of pywbem did not upgrade to our latest version due to new dependencies (nocasedict and nocaselist). In the pywbemtools project I think we would always support the same Python versions as pywbem does, for consistency.

The question is at which point do we drop support for Python 2.7. Maybe we should set 6/2024 as a date for that.

**Comment by GhostOps77** [comment]
@andy-maier The drop for Python 2 support is up to you and to @KSchopmeyer
So, If you guys still want the Py2 support, do you guys mind in supporting in improving the code in [this](https://github.com/GhostOps77/click-repl/tree/GhostOps77-patch-1) fork? (I can't make the code work for python 2, I really need some help)

But my question is, If you guys are going to do the updates in click-repl, where can I contact you guys?

**Comment by andy-maier** [comment]
Hi @GhostOps77 . Thanks for the offer. We'll discuss it in our weekly meeting tomorrow. We have a public Slack channel for pywbem development at https://pywbem.slack.com/archives/CA2516Z0Q . You're welcome to join there (I think you can just go there - if that does not work, let me know here).

**Comment by GhostOps77** [comment]
@andy-maier good, can u add me in there? it says im not tied in into that workspace
add me in there please
(will delete this comment after i've been added)
email: muhsen.kaamil2003@gmail.com

**Comment by andy-maier** [comment]
I invited you just now.

**Comment by andy-maier** [comment]
BTW, Karl is travelling this week, so we canceled our weekly meeting this week.

**Comment by KSchopmeyer** [comment]
Status of this issue:

1. The current release of click-repl does not effectively support tab completion so that use of tab completion in the interactive mode does not really work.
2. Code is complete for most of the required changes but we do not know what the release date is for the next version of click-repl and do not want to wait for that release.
3. There is still one feature missing from click-repl (tab-completion for directories and files) and that is being worked on by another contributor.

Thus we propose to release pywbemtools with the current (0/2.0) version of click-repl and the statement in the documentation that the tab-completion is incomplete.  We will rerelease upon release of an updated version of click-repl.

**Comment by andy-maier** [comment]
The remaining work for us is to require the new click-repl version once it is released. We keep this issue in 1.3.0 for the time being, to see whether a new click-repl version is released by the time pywbemtools 1.3.0 is ready for release.

**Comment by KSchopmeyer** [comment]
ACTIONS: 

1. Clarify what versions of the click-repl code we can use
2. clarify what limitations are in the repl mode.

**Comment by KSchopmeyer** [comment]
Move this issue to version 1.4.0

**Comment by KSchopmeyer** [comment]
click-repl 3.0 has been released however, the functionality to make the general options available to the repl command line has disappeared.  Testing with click-repl 3.0 causes over 80 test failures in the unit pywbemcli tests.

**Comment by andy-maier** [comment]
This issue is now handled in issue #1565 . Closing this issue here.
